billing: Downgrade plan to Limited during realm deactivation.

hackerkid opened the pull request.

timabbott commented:
The code changes look reasonable, but can you add comments documenting how this works more clearly in `downgrade_for_realm_deactivation`?  

Also, we probably need an `if settings.BILLING_ENABLED` conditional inside `do_deactivate_realm`; I think this change as written will crash if that setting is disabled (please test it locally!).

hackerkid commented:
@timabbott This is ready for a review. The travis tests have passed but for some reason GitHub shows its still running. 

> Also, we probably need an if settings.BILLING_ENABLED conditional inside do_deactivate_realm; I think this change as written will crash if that setting is disabled (please test it locally!).

Sorry about that. I completely forgot about the BILLING_ENABLED conditional. I have updated the commit. Without it, realm deactivation flow would have errored if BILLING_IS_DISABLED.

Since you last reviewed, I made some additional changes to PR.

* Adding some helper functions.
* Adding tests for an existing helper function.
* Updated the commit which fixed the bug in /billing page. The commit I  added initially fixed the issue by redirecting /billing to /upgrade page for realms with expired plans. I realized that is not the right approach. So I instead fixed the bug for now https://github.com/zulip/zulip/pull/14224/commits/35c7bfe4ffc69c2beb7bf131b6f0691435e9a89d. A proper follow-up would be to not show licenses and other info in /billing page if the realm is on Limited plan.

I have tested this PR by
* Creating a realm and deactivating it and then accessing the /billing page to make sure that the plan is shown as Free.
* Creating a new realm and deactivating it while BILLING_ENABLED is False and make sure that nothing goes wrong.
* Tests in test_stripe.py to make sure that we do the right thing during realm is deactivated.
* Also ran all the stripe_tests with --generate-fixtures enabled to make sure that nothing is broken.

What I am thinking on working next.

* Followup of the last commit https://github.com/zulip/zulip/pull/14224/commits/35c7bfe4ffc69c2beb7bf131b6f0691435e9a89d as discussed above.
* I am thinking of starting to work on https://github.com/zulip/zulip/pull/13898. Atlast the backend part of it. Its probably easier for me to finish it now than having to do it later.

timabbott commented:
Looks great, merged, thanks @hackerkid!  And I agree with those next projects :)
